fe/fe-core/src/main/java/com/starrocks/external/hive/events/MetastoreEventsProcessor.java

```diff
@@ -0,0 +1,408 @@
+     *               and then returned using this. Otherwise, all the events are returned.
+     * @return List of NotificationEvents from metastore since eventId.
+     * @throws MetastoreNotificationFetchException In case of exceptions from HMS.
+     */
+    private List<NotificationEvent> getNextHMSEvents(String resourceName,
+                                                    final boolean getAllEvents,
+                                                    @Nullable final IMetaStoreClient.NotificationFilter filter)
+            throws MetastoreNotificationFetchException {
+        Long lastSyncedEventId = null;
+        try {
+            HiveMetaClient client = hiveRepository.getClient(resourceName);
+            if (client == null) {
+                return Collections.emptyList();
+            }
+
+            lastSyncedEventId = lastSyncedEventIds.get(resourceName);
+            // restart fe or just created hive table.
+            if (lastSyncedEventId == null) {
+                lastSyncedEventIds.put(resourceName, client.getBaseHmsEventId());
+                return Collections.emptyList();
+            }
+
+            CurrentNotificationEventId currentNotificationEventId = client.getCurrentNotificationEventId();
+            long currentEventId = currentNotificationEventId.getEventId();
+            if (currentEventId == lastSyncedEventId) {
+                return Collections.emptyList();
+            }
+
+            int batchSize = getAllEvents ? -1 : Config.hms_events_batch_size_per_rpc;
+            NotificationEventResponse response = client.getNextNotification(lastSyncedEventId, batchSize, null);
+            LOG.info(String.format("Received %d events. Start event id : %d",
+                    response.getEvents().size(), lastSyncedEventId));
+
+            if (filter == null) {
+                return response.getEvents();
+            }
+
+            List<NotificationEvent> filteredEvents = new ArrayList<>();
+            for (NotificationEvent event : response.getEvents()) {
+                if (filter.accept(event)) {
+                    filteredEvents.add(event);
+                }
+            }
+
+            return filteredEvents;
+        } catch (DdlException e) {
+            throw new MetastoreNotificationFetchException(
+                    "Unable to fetch notifications from metastore. Last synced event id is "
+                    + lastSyncedEventId, e);
+        }
+    }
+
+    /**
+     * Fetch the next batch of NotificationEvents from metastore. The default batch size is
+     * <code>{@link Config#hms_events_batch_size_per_rpc}</code>
+     */
+    private List<NotificationEvent> getNextHMSEvents(String resourceName)
+            throws MetastoreNotificationFetchException {
+        return getNextHMSEvents(resourceName, false, null);
+    }
+
+    private void doExecuteWithPartialProgress(List<MetastoreEvent> events) {
```

Comment:
Can this method be merged with doExecute with Parallelism setting to 1.

Reply:
I think it is more intuitive to judge according to the bool configuration. Single-threaded execution is mainly for debugging if there is a bug.